### ISSUE TITLE
Feature request: Determine Python project root & venv by looking for pyproject.toml (aka monorepo support)

### ISSUE THREAD

**Issue by codethief** [microsoft/pyright#10498]
### Current behavior
Currently, with a workspace folder structure like

```
workspace-root
|- .git/
|- projectA/
|  |- .venv/
|  |- somePackage/
|  |- .python-version
|  |- main.py
|  |- pyproject.toml
|  |- uv.lock
|- … (maybe other projects)
```

Pylance will

1. fail to find the .venv automatically (i.e. will fail to set `python.pythonPath` automatically), and
2. generate import suggestions for symbols from `somePackage` that are prefixed with `projectA.somePackage.`, even they should begin with only `somePackage.` This is because "[the] language server treats the workspace root (i.e. folder you have opened) as the main root of user module imports" ([source](https://github.com/microsoft/pylance-release/blob/main/TROUBLESHOOTING.md#unresolved-import-warnings)).

### Known workarounds

- https://marketplace.visualstudio.com/items?itemName=teticio.python-envy for automatically changing the venv depending on the file that's open.
- Multi-root workspaces, as per https://github.com/microsoft/pylance-release/issues/3899#issuecomment-1414112654 and https://github.com/microsoft/vscode-python/wiki/Mono-Repo-Set%E2%80%90up-Guide .
- Adjusting `python.analysis.extraPaths` to fix imports, compare
https://github.com/microsoft/pylance-release/blob/main/TROUBLESHOOTING.md#unresolved-import-warnings

Unfortunately, for larger monorepos with a more complicated folder structure the workarounds get unwieldy quickly: 
- Multi-root workspaces make it difficult to discover and navigate the actual, full directory structure of the monorepo. In particular, project boundaries are not necessarily always determined by *Python* project boundaries. EDIT: https://github.com/microsoft/pylance-release/issues/5564 lists many more reasons why multi-root workspaces are not a good solution. 
- `python.analysis.extraPaths` would have to be set on a per-project basis (as opposed to workspace-wide / monorepo-wide) to prevent accidental cross-imports, e.g. through different [execution environments](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#execution-environment-options). Either way, `extraPaths` needs to be set in a `pyrightconfig.json` or a `pyproject.toml` in the workspace *root*, nowhere else, see also https://github.com/microsoft/pylance-release/issues/5564.

### Suggested behavior

Pylance should automatically find both, the venv "responsible" for a given file and the root of user module imports, based on the location of the nearest `pyproject.toml` file. (See https://github.com/microsoft/pylance-release/issues/5564 for determining the nearest `pyproject.toml`.) Now, there is a bit of vagueness here, because
- venvs can be stored anywhere, including outside the workspace folder, 
- there is not *one* folder structure that all Python projects follow, `sys.path` can be modified by hand, etc. etc.

However, I would argue a decent-enough 80/20 solution would be to
- automatically use venvs that are placed alongside the `pyproject.toml` in a `.venv` folder (i.e. the default structure used by `uv`).
- support [flat-layout and src-layout directory structures](https://setuptools.pypa.io/en/latest/userguide/package_discovery.html#src-layout), i.e. either treat the parent folder of `pyproject.toml` as main root of user module imports or a given subfolder thereof (like `src`). Possibly add a setting to `[tool.pyright]` to distinguish these cases.

I assume even this approach is probably still not without challenges, in particular there seems to be a [dependency on the core Python extension](https://github.com/microsoft/pylance-release/issues/3899#issuecomment-1414106653) which I don't know much about. However, I thought I'd still at least bring up this issue – many aspects of it have been discussed here on GitHub over time but none of them has been solved satisfactorily I'd say and it keeps on biting me. Either way, thank you for considering this and thank you for your time!

### Related tickets
- https://github.com/microsoft/pylance-release/issues/3899 (nested projects)
- https://github.com/microsoft/pylance-release/issues/5564
- https://github.com/microsoft/pylance-release/issues/7151

Outside microsoft/pylance-release: 
- https://github.com/microsoft/vscode-python/issues/21204
- https://github.com/astral-sh/uv/issues/9637#issuecomment-2899363726

**Comment by erictraut** [comment]
This feature request would require _very_ significant work and restructuring of pyright's logic to support. It would also require significant feature spec work to make sure it composes well with existing venv discovery mechanisms. I understand the desire for this feature, but it would be difficult to justify all of the work involved unless there's strong signal from many pyright users that this is an important feature. 

Pyright already has support for something called "execution environments". It was designed to handle monorepos. Have you tried this mechanism? Does it (partially) address your use case?

I'm going to close the feature request for now because I don't have any current plans to implement this, but I might consider it in the future if the issue receives sufficient thumbs-up reactions.

**Comment by codethief** [comment]
@erictraut Thanks so much for the feedback!

> Pyright already has support for something called "execution environments". It was designed to handle monorepos. Have you tried this mechanism? Does it (partially) address your use case?

As a workaround, maybe. What I find rather annoying about execution environments is:

- They have to be configured in a config file in the workspace / monorepo root, even if e.g. I only use Python in some of my projects deep down the folder hierarchy. It's also awkward from the POV of code ownership if different teams own different projects/folders.
- Configuring the execution environments is manual work that could be automated away. (pythonVersion is given by the venvs of the individuals projects, the root path could be derived from the presence of pyproject.toml & .venv in the subfolders, …)
- I pin the Pyright version per project, not for the whole monorepo. So it doesn't make much sense to me to define a repo-global config with different execution environments for Pyright when I actually want to keep the projects in my monorepo separate and not tie their type checking together in any way.

Maybe most of this could already be solved by having Pyright determine "execution environments" automatically by looking for `pyproject.toml` files?

**Comment by codethief** [comment]
For anyone else coming here (and future me), the standard way I've worked around the issue in various Python monorepos since my last post is the following: 

Assumptions: 
- I have a monorepo of multiple Python projects, each with its own `pyproject.toml` and virtual environment. In VSCode the monorepo is a single VSCode workspace.
- In VSCode I use Pylance together with the [Python Envy](https://marketplace.visualstudio.com/items?itemName=teticio.python-envy) extension to switch between the venvs of the different Python projects automatically as I switch between files/tabs.

Then in the repo root I create a file `pyrightconfig.json` with the following contents:

```javascript
{
    // This file serves to configure *only* the Pylance language server in
    // VSCode and work around
    // https://github.com/microsoft/pylance-release/issues/7252.
    //
    // Meanwhile, when running Pyright from the command line in a given Python
    // project, the project folder should contain another pyrightconfig.json, so that
    // Pyright won't pull in the present file and type-check the entire
    // repository at once.

    "executionEnvironments": [
        // We don't set pythonVersion etc. for the individual execution
        // environments because the python-envy extension for VSCode should set
        // the venv automatically.
        {
            "root": "./project_a"
        },
        {
            "root": "./project_b"
        }
    ],

    // WARNING WARNING WARNING: Keep these in sync with each project's pyrightconfig.json!
    "enableExperimentalFeatures": true, // We want PEP 764 support
    "strictDictionaryInference": true,
    "strictListInference": true,
    "strictSetInference": true,
}
```

Then in each project folder (here: `./project_a`, `./project_b`) I create additional `pyrightconfig.json` files:

```javascript 
{
    // This file is ignored by the Pylance VSCode extension.
    // See /pyrightconfig.json in the repo root before modifying this file!

    "venv": ".venv",
    "venvPath": ".",

    // WARNING WARNING WARNING: Keep these in sync with the root /pyrightconfig.json!
    "enableExperimentalFeatures": true, // We want PEP 764 support
    "strictDictionaryInference": true,
    "strictListInference": true,
    "strictSetInference": true,
}
```

This way,
- import suggestions in VSCode work as expected when editing files in any of the projects, and
- imports can be resolved in both Pyright (`cd project_(a|b) && uv run pyright`) and Pylance/VSCode. They can be made to work even with _editable_ cross-project dependencies, e.g. `project_b` referencing `project_a` in its `pyproject.toml` like so: `[tool.uv.sources] project_b = { path = "../project_a", editable = true }`. The trick is to [avoid using setuptools in `project_a` and using, e.g., hatchling instead](https://github.com/microsoft/pyright/issues/3846#issuecomment-1227817419).
